src/CSET/_workflow_utils/fetch_data.py

```diff
@@ -0,0 +1,202 @@
+        "forecast_length": isodate.parse_duration(os.environ["CSET_ANALYSIS_PERIOD"]),
+        "forecast_offset": isodate.parse_duration(os.environ["CSET_ANALYSIS_OFFSET"]),
+        "share_dir": os.environ["CYLC_WORKFLOW_SHARE_DIR"],
+        "cycle_point": os.environ["CYLC_TASK_CYCLE_POINT"],
+        "model_number": os.environ["MODEL_NUMBER"],
+    }
+    try:
+        variables["data_period"] = isodate.parse_duration(os.environ["DATA_PERIOD"])
+    except KeyError:
+        # Data period is not needed for initiation time.
+        if variables["date_type"] != "initiation":
+            raise
+        variables["data_period"] = None
+    return variables
+
+
+def _template_file_path(
+    raw_path: str,
+    date_type: Literal["validity", "initiation", "lead"],
+    data_time: datetime,
+    forecast_length: timedelta,
+    forecast_offset: timedelta,
+    data_period: timedelta,
+) -> list[str]:
+    """Fill time placeholders to generate a file path to fetch."""
+    placeholder_times: list[datetime] = []
+    lead_times: list[timedelta] = []
+    match date_type:
+        case "validity":
+            date = data_time
+            while date < data_time + forecast_length:
+                placeholder_times.append(date)
+                date += data_period
+        case "initiation":
+            placeholder_times.append(data_time)
+        case "lead":
+            placeholder_times.append(data_time)
+            lead_time = forecast_offset
+            while lead_time < forecast_length:
+                lead_times.append(lead_time)
+                lead_time += data_period
+        case _:
+            raise ValueError(f"Invalid date type: {date_type}")
+
+    paths: list[str] = []
+    for placeholder_time in placeholder_times:
+        # Expand out all other format strings.
+        path = placeholder_time.strftime(os.path.expandvars(raw_path))
+        if lead_times:
+            # Expand out lead time format strings, %N.
+            for lead_time in lead_times:
+                # BUG: Will not respect escaped % signs, e.g: %%N.
+                paths.append(
+                    path.replace("%N", f"{int(lead_time.total_seconds()) // 3600:03d}")
+                )
+        else:
+            paths.append(path)
+    return paths
+
+
+def fetch_data(file_retriever: FileRetriever = FilesystemFileRetriever):
+    """Fetch the data for a model.
+
+    The following environment variables need to be set:
+    * CSET_ANALYSIS_OFFSET
+    * CSET_ANALYSIS_PERIOD
+    * CYLC_TASK_CYCLE_POINT
+    * CYLC_WORKFLOW_SHARE_DIR
+    * DATA_PATH
+    * DATA_PERIOD
+    * DATE_TYPE
+    * MODEL_NUMBER
+
+    Parameters
+    ----------
+    file_retriever: FileRetriever
+        FileRetriever implementation to use. Defaults to FilesystemFileRetriever.
+    """
+    v = _get_needed_environment_variables()
+
+    # Prepare output directory.
+    cycle_share_data_dir = (
+        f"{v['share_dir']}/cycle/{v['cycle_point']}/data/{v['model_number']}"
```

Comment:
[Not a cylc provided one](https://cylc.github.io/cylc-doc/stable/html/reference/job-script-vars/). Note that this is within the share directory, rather than being per job, as we need this data to be discoverable by later tasks.

Reply:
Just checked the docs, I was thinking of `ROSE_DATAC` which is `~/cylc-run/CYLC_WORKFLOW_ID/share/cycle/CYLC_TASK_CYCLE_POINT` at cylc 8

Reply:
Switched to `ROSE_DATAC`.

Reply:
I don't think it changes that we want to do this, but from a near future version of cylc (https://github.com/cylc/cylc-flow/pull/6117) there will be a `CYLC_SHARE_CYCLE_DIR` variable that does the same thing as `ROSE_DATAC`, and it'll be [recommended to migrate](https://github.com/metomi/rose/pull/2782). I think we can keep as is for now, as we will have to wait a bit for everyone's cylc to be updated sufficiently.